src/org/labkey/test/tests/SampleSetRemoteAPITest.java

```diff
@@ -483,39 +487,46 @@ private AssayDesignerPage generateAssay(String subfolderName, String assayName)
 
         TestDataGenerator dgen2 = new TestDataGenerator(assayLookup)
                 .withColumnSet(resultsFieldset)
                 .addCustomRow(Map.of("ParticipantID", "Harry", "Date", "10/11/2018", "SampleName", "Green", "SampleVolume", 12.5))
                 .addCustomRow(Map.of("ParticipantID", "William", "Date", "10/12/2018", "SampleName", "Red", "SampleVolume", 14.5))
                 .addCustomRow(Map.of("ParticipantID", "Jenny", "Date", "10/13/2018", "SampleName", "Yellow", "SampleVolume", 17.5))
                 .addCustomRow(Map.of("ParticipantID", "Hermione", "Date", "10/14/2018", "SampleName", "Orange", "SampleVolume", 11.5));
-        String pasteData2 = dgen2.writeTsvContents();
 
         TestDataGenerator dgen3 = new TestDataGenerator(assayLookup)
                 .withColumnSet(resultsFieldset)
                 .addCustomRow(Map.of("ParticipantID", "George", "Date", "10/11/2018", "SampleName", "Green", "SampleVolume", 12.5))
                 .addCustomRow(Map.of("ParticipantID", "Arthur", "Date", "10/12/2018", "SampleName", "Red", "SampleVolume", 14.5))
                 .addCustomRow(Map.of("ParticipantID", "Colin", "Date", "10/13/2018", "SampleName", "Yellow", "SampleVolume", 17.5))
                 .addCustomRow(Map.of("ParticipantID", "Ronald", "Date", "10/14/2018", "SampleName", "Orange", "SampleVolume", 11.5));
-        String pasteData3 = dgen3.writeTsvContents();
 
+        return Arrays.asList(dgen1, dgen2, dgen3);
+    }
+
+    private void insertAssayData(String assayName, List<TestDataGenerator> dataGenerators)
+    {
+        // assumes we are at the assay list
         waitAndClickAndWait(Locator.linkWithText(assayName));
         AssayRunsPage assayRunsPage = new AssayRunsPage(getDriver());
         DataRegionTable runsTable = assayRunsPage.getTable();
         runsTable.clickHeaderButton("Import Data");
         clickButton("Next");
 
-        // insert 3 runs
-        new AssayImportPage(getDriver()).setNamedTextAreaValue("TextAreaDataCollector.textArea", pasteData1);
-        clickButton("Save and Import Another Run");
-        new AssayImportPage(getDriver()).setNamedTextAreaValue("TextAreaDataCollector.textArea", pasteData2);
-        clickButton("Save and Import Another Run");
-        new AssayImportPage(getDriver()).setNamedTextAreaValue("TextAreaDataCollector.textArea", pasteData3);
-        clickButton("Save and Finish");
-
-        return Arrays.asList(dgen1, dgen2, dgen3);
+        int limit = dataGenerators.size();
+        int imported = 0;
+        for(TestDataGenerator dataGen : dataGenerators)
+        {
+            new AssayImportPage(getDriver()).setNamedTextAreaValue("TextAreaDataCollector.textArea",
+                    dataGen.writeTsvContents());
+            imported++;
+
+            if(imported < limit)
+                clickButton("Save and Import Another Run");
+            else
+                clickButton("Save and Finish");
```

Comment:
`AssayImportPage` has methods for clicking these save buttons

Reply:
refactored to use the page methods